src/plugins/intel_gpu/src/graph/impls/ocl/gemm.cpp

```diff
@@ -95,10 +113,28 @@ struct gemm_impl : multi_stage_primitive<gemm> {
         for (size_t s = 0; s < stage; s++) {
             kernel_offset += _kernels_data[s].kernels.size();
         }
-        for (size_t kd_idx = 0; kd_idx < _kernels_data[stage].kernels.size(); ++kd_idx) {
-            if (_kernels_data[stage].kernels[kd_idx].skip_execution)
-                continue;
+        size_t kd_idx = 0;
+        if (is_dynamic()) {
+            bool not_divisible_k = true;
+            bool not_divisible_n = true;
+            auto prim_params = std::dynamic_pointer_cast<kernel_selector::gemm_params>(_kernels_data[stage].params);
+            auto _shape_info_memory = instance.shape_info_memory_ptr();
+            mem_lock<int32_t> lock(_shape_info_memory, stream);
+            auto shape_info_ptr = lock.data();
+            not_divisible_k = evaluateJIT(prim_params->not_divisible_k, shape_info_ptr);
+            not_divisible_n = evaluateJIT(prim_params->not_divisible_n, shape_info_ptr);
+            if (not_divisible_k == false && not_divisible_n == false) {
+                kd_idx = 0;
+            } else if (not_divisible_k == false && not_divisible_n == true) {
+                kd_idx = 1;
+            } else if (not_divisible_k == true && not_divisible_n == false) {
+                kd_idx = 2;
+            } else if (not_divisible_k == true && not_divisible_n == true) {
+                kd_idx = 3;
+            }
```

Comment:
Umm couldn't we just do this kernel selection at update_dispatch_data, (as it is being done for fc), where we can easily detect the size. There we can set the unexpected kernel to be skept.

Reply:
I'll update it as you reviewed. Thank you!

Reply:
I moved the kernel selection logic to `GetUpdateDispatchDataFunc`, and removed the use of `mem_lock`.